### ISSUE TITLE
Ingredient parsing may swallow parse errors

### ISSUE THREAD

**Issue by Shadows-of-Fire** [neoforged/NeoForge#743]
I was porting something containing the following Ingredient:

```json
{
    "type": "forge:nbt",
    "item": "minecraft:potion",
    "nbt": "{\"Potion\": \"minecraft:night_vision\"}"
}
```

Since the ingredient type is now `neoforged:nbt`, I would expect this to error.  However, this actually results in a successful deserialization with an ingredient containing just the `minecraft:potion` item.

**Comment by Technici4n** [comment]
That's a side-effect of using Either codecs. We probably want a way to skip the second codec if some key exists in the codec.

**Comment by KnightMiner** [comment]
Didn't we discuss this issue on Discord before? The solution recommend was instead of using an either codec, to just have the type field default to the ID of the vanilla serializer in the forge registry.

**Comment by marchermans** [comment]
Yes this was a direct concequence of the requirement that if you used the vanilla ingredient it should not have a "type" field.

If somebody knows a better codec setup to do this, I would be open to figure this out.

**Comment by FiniteReality** [comment]
> If somebody knows a better codec setup to do this, I would be open to figure this out.

My thought would be a custom MapCodec based on KeyDispatchCodec which replaces the typeKey/keyCodec fields with a MapCodec which is built with optionalFieldOf

**Comment by Shadows-of-Fire** [comment]
I have [this codec](https://github.com/Shadows-of-Fire/Placebo/blob/1.20.4/src/main/java/dev/shadowsoffire/placebo/codec/MapBackedCodec.java) which is a dispatch codec with the ability to set a default type

**Comment by Matyrobbrt** [comment]
Fixed in [7b6ac97](https://github.com/neoforged/NeoForge/commit/7b6ac97857922b92bda7dc5461530e916e1e9a99)
